### ISSUE TITLE
Add override/extends information to Decorations [phab:D1009]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1345]
_Created by **zarko** at 2016-08-13 00:26:10_: 

It's useful to have information about the overrides/extends
effects of binding sites inline with file decorations
(e.g. to show indicators in a UI that link to the thing
being extended).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

Maybe do this before checking the error, since the latency is still interesting in that case.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

Move below the check, please.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

Suggestion:
```
pb := &xpb.DecorationsReply_Overrides{
   Override: make([]*xpb.DecorationsReply_Override, len(eos)),
}
for i, eo := range eos {
  pb.Override[i] = eo
  extendOverridesTargets.Add(eo.Ticket)
}
```

Or, if you prefer the append, just drop the `make` call entirely and say `pb := new(xpb.DecorationsReply_Overrides)`.

It is usually also safe to alias within protos if you're about to write back a result anyway, but this function is too long for me to be confident that won't cause problems.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

You don't need to check for nil-ness here; iteration and reads are fine on a nil map (they return zeroes or no items, respectively).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

As line 660.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

`if extendsOverridesTargets.Contains(tgt) {`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-13 00:48:14_:

Not sure it matters, but is `OVERRIDES` the more sensible default here?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-08-15 17:27:34_:

If `extendsOverridesTargets` is nil, we can avoid the allocation of `request` (since we'll never add any tickets to it).

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-15 17:28:06_:

  - Address comments.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-08-15 17:37:59_:

Actually, now that I look, I realize it will never be nil anyway. The semantic you probably want are `len(extendsOverrideTargets) != 0`

Because `stringset.New` always allocates a map. I am thinking about replacing this with bitbucket.org/creachadair/stringset which avoids some of those little corners.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-15 17:43:08_:

  - Address comments.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2016-08-15 17:48:35_

**Comment by kythe** [comment]
_**zarko** closed via commit 3ef226d848add9bef9dfd78d0f11c8c9044d4953 at 2016-08-15 17:51:18_
